File: coding/solutions/src/test/java/blind75/TwoSumTest.java
package blind75;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import xyz.grind.coding.blind75.TwoSum;

import static org.junit.jupiter.api.Assertions.assertAll;
import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertNull;

public class TwoSumTest {
    @Test
    @DisplayName("Should return array that contains two elements that sum up to target uf found these elements")
    void shouldReturnArrayWithTwoIntegersThatSumUpToTarget() {
        assertAll(
            () -> assertArrayEquals(TwoSum.twoSumBruteForce(new int[]{20,2,3,8,5,3}, 28), new int[]{0, 3}),
            () -> assertArrayEquals(TwoSum.twoSumBruteForce(new int[]{2,7,11,15}, 9), new int[]{0,1}),
            () -> assertArrayEquals(TwoSum.twoSumBruteForceTwo(new int[]{3,2,4}, 6), new int[]{1,2}),
            () -> assertArrayEquals(TwoSum.twoSumBruteForceTwo(new int[]{3,1,20,89}, 90), new int[]{1,3}),
            () -> assertArrayEquals(TwoSum.twoSum(new int[]{20,2,3,8,5,3}, 28), new int[]{0, 3}),
            () -> assertArrayEquals(TwoSum.twoSum(new int[]{2,7,11,15}, 9), new int[]{0,1}),
            () -> assertArrayEquals(TwoSum.twoSum(new int[]{3,2,4}, 6), new int[]{1,2}),
            () -> assertArrayEquals(TwoSum.twoSum(new int[]{3,1,20,89}, 90), new int[]{1,3})
        );
    }

    @Test
    @DisplayName("Should return null if no the array doesn't contain two numbers that sum up to target")
    void shouldReturnEmptyArrayIfNoTwoIntegersSumUpToTarget() {
        assertAll(
            () -> assertNull(TwoSum.twoSum(new int[]{20,2,3,8,5,3}, 1324)),
            () -> assertNull(TwoSum.twoSum(new int[]{2,7,11,15}, 4332)),
            () -> assertNull(TwoSum.twoSum(new int[]{3,2,4}, 87)),
            () -> assertNull(TwoSum.twoSum(new int[]{3,1,20,89}, 2434))
        );
    }

}

File: problem-solving/solutions/leetcode/blind75/src/test/kotlin/easy/TwoSumTest.kt
package easy

import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Assertions.assertArrayEquals
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.function.Executable
import xyz.grind.coding.easy.TwoSum

class TwoSumTest {
    @Test
    fun `Should return array that contains two elements that sum up to target uf found these elements`() {
        Assertions.assertAll(
            Executable {
                assertArrayEquals(TwoSum.twoSumBruteForce(intArrayOf(20, 2, 3, 8, 5, 3), 28), intArrayOf(0, 3))
            },
            Executable {
                assertArrayEquals(TwoSum.twoSumBruteForce(intArrayOf(2, 7, 11, 15), 9), intArrayOf(0, 1))
            },
            Executable {
                assertArrayEquals(TwoSum.twoSumBruteForceTwo(intArrayOf(3, 2, 4), 6), intArrayOf(1, 2))
            },
            Executable {
                assertArrayEquals(TwoSum.twoSumBruteForceTwo(intArrayOf(3, 1, 20, 89), 90), intArrayOf(1, 3))
            },
            Executable {
                assertArrayEquals(TwoSum.twoSum(intArrayOf(20, 2, 3, 8, 5, 3), 28), intArrayOf(0, 3))
            },
            Executable {
                assertArrayEquals(TwoSum.twoSum(intArrayOf(2, 7, 11, 15), 9), intArrayOf(0, 1))
            },
            Executable {
                assertArrayEquals(TwoSum.twoSum(intArrayOf(3, 2, 4), 6), intArrayOf(1, 2))
            },
            Executable {
                assertArrayEquals(TwoSum.twoSum(intArrayOf(3, 1, 20, 60), 80), intArrayOf(2, 3))
            }
        )
    }

    @Test
    fun `Should return zero if no the array doesn't contain two numbers that sum up to target`() {
        Assertions.assertAll(
            Executable {
                assertEquals(TwoSum.twoSum(intArrayOf(20, 2, 3, 8, 5, 3), 1324)[0], 0)
            },
            Executable {
                assertEquals(TwoSum.twoSum(intArrayOf(2, 7, 11, 15), 4332)[0], 0)
            },
            Executable {
                assertEquals(TwoSum.twoSum(intArrayOf(3, 2, 4), 87)[0], 0)
            },
            Executable {
                assertEquals(TwoSum.twoSum(intArrayOf(3, 1, 20, 89), 2434)[0], 0)
            }
        )
    }

    @Test
    fun `Should check handling edge cases`() {
        Assertions.assertAll(
            Executable {
                assertArrayEquals(TwoSum.twoSumBruteForce(intArrayOf(), 0), intArrayOf())
            },
            Executable {
                assertArrayEquals(TwoSum.twoSumBruteForceTwo(intArrayOf(), 0), intArrayOf())
            }
        )
    }
}
